apps/newsletters-ui/src/app/components/StepNav.tsx

```diff
@@ -31,6 +84,31 @@ export const StepNav = ({
 		}
 	});
 
+	const updateCompletion = () => {
+		const list = stepperConfig.steps.reduce<
+			Partial<Record<string, boolean | undefined>>
+		>((record, step) => {
+			const result = step.schema
+				? step.schema.safeParse(formData).success
+				: undefined;
+
+			return { ...record, [step.id]: result };
+		}, {});
+
+		setCompletionRecord(list);
+	};
+
+	// On the initial render, the completionRecord is set to {}
+	if (!Object.keys(completionRecord).length) {
+		updateCompletion();
+	}
+
```

Comment:
little tip I got taught at a previous company that had a typedoc only policy...
It's quite nice and I use it quite a bit.

const isCompletionRecordPopulated = !Object.keys(completionRecord).length;
if (isCompletionRecordPopulated) {
  updateCompletion();
}
This is not a request to change, just sharing something I learned from someone smarter than me.